### DISCUSSION TITLE
securing the bluetooth pairing with pin code?

### DISCUSSION THREAD

**Discussion by DariBer** [sle118/squeezelite-esp32#217] [Ideas]
I have a question regarding securing Bluetooth pairing a device like iPhone with a Squeezelite-ESP32 . When pairing an iPhone or what ever using bluetooth, do you know if I can secure that process with maybe a 4 digit code? Cause now anyone who’s near (like on the street, neighbours in the house) can connect to it and start to play music, as it doesn’t need any code. So is there a chance to set a pin code and only the one who know that code can pair and start use the device for streaming?

**Comment by sle118** [comment]
This has been on my to-do list for a while. Il move that to a new issue to keep it on the radar.

**Reply by DariBer** [reply]
It would be really nice to have that security option, so if you implement it, thanx in advance :-)

**Comment by eigenphase** [comment]
Yes, please!!

**Comment by sle118** [comment]
I've managed to fix this in my local development environment, but this change will be extremely disruptive for our user base. We would need to move away from the current security protocol, back to the legacy "pin code" approach. Since we don't collect usage information, I can only imagine that the change might disrupt the daily use of the majority of our users. Unfortunately, it is not possible to make this an option either. Either we roll back to legacy pin code and cause some pain to existing users, or I need to find another "clever" way around this which will never really be complete.  I might try to poll our user base to see what they think in the logitech forums and decide the path forward from there.

**Reply by eigenphase** [reply]
This is just for my understanding, so please don't take my question the wrong way. But how is there a "current security protocol" if there is no security at all? In other words, if any arbitrary Bluetooth adapter can connect?

I totally get the concern with disrupting current installations but I'm wondering if there is any path to support both. May at least a BT MAC address filter? Pre-processor directive to select the protocol? Or even better a runtime option to select the protocol (in the same way as Bluetooth can be enabled/disabled)

**Reply by sle118** [reply]
I wish someone can prove me wrong, but at compile time, is either SSP (no security because there's no input or output supported by default on the hardware) or legacy pin code.

Although I'm doing my best to support the project, any feature implementation requires a substantial amount of work. Yes, I could design a solution that would keep SSP active and not be disruptive. But to achieve this, I would need to build a UI to support interactions/actions and a UI needs backend service support to work. So assuming I have a design, I need to implement that.... New code in a partition that is full already so now I need to start looking for a few kilobytes to shave somewhere else. Then once everything fits, I have to start a round of tests to ensure I didn't break anything and that the new logic works.

Now keep in mind that this software is maintained "for fun" and when time permits and the quickest path to BT security is to go legacy. 

We encourage anyone wanting to share their enhancement to our code base to submit pull requests with their changes. I can think of many ways to better support this requirement that other users have, but none that I could achieve with a level of effort that would not impact my personal life as I also need to work full time to earn a living.

In short, everything is possible technically, but not humanly.

Thank you for your understanding!

**Reply by DariBer** [reply]
@sle118 I just wonder if you have any update about this, is there any way that this some how been released, either with the PIN code or any other stuff?

I really want to start use this but now the BT integration is a show stopper

**Reply by emlynmac** [reply]
@sle118 as this is such a large break for the mainstream, is it something that could be left on a branch for those to build/ install / patch as needed?

I am interested in getting into the software side, but finding time to learn the esp tools is currently tricky...

**Reply by sle118** [reply]
I have been extremely busy in recent months, with work often overlapping with spare time so unfortunately there wasn't much left to address this. 

@emlynmac I was planning on implementing a "pair mode" as an option that allows pairing for a limited time, with clients being stored in a white list of mac addresses. But this is no small feature. I'm sure I can handle the bulk of this through command line (long ago I built a mechanism to surface these automatically in the UI). So what is needed is enabled pairing command (timed out Airways), then another command to remove some white listed devices. 

I just don't know when I manage to get to this.

**Reply by emlynmac** [reply]
Is there a doc or something for where parts of the code are? I'd love to get more into this

**Reply by emlynmac** [reply]
@sle118 could we also make this a build option so those who want can use a different version?

**Reply by sle118** [reply]
see https://github.com/sle118/squeezelite-esp32/discussions/217#discussioncomment-8060479

I will see if I can manage to do something in the current version, but I simply don't have any bandwidth right now.

**Reply by emlynmac** [reply]
@sle118 I read that, hence the offer to take that on further. You had written previously that you had it working locally, so that would be a point for me to start looking at.

**Comment by seancmalloy** [comment]
Any update on implementing secure bluetooth connection with pin? I have multiple networked speakers with squeezelite-esp32 in my condo and it's extremely annoying when someone connects it them but I want to keep the  bluetooth option.

**Comment by sle118** [comment]
As stated in https://github.com/sle118/squeezelite-esp32/discussions/342, the current version will be limited to bug fixes as the bulk of my spare time is dedicated to the refactoring of the system. Implementing changes in the current version had become very annoying as we have to fight for every byte on the flash storage, so it is unlikely that I will implement this feature in the old version.

It is still not clear to me if I'll be able to provide a version that fits in 4Mb of flash, but so far it seems that a limited version might be possible for devices with smaller flash chips.
